File: test/AuthZI.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraExternalID.NET9.0/Initialization/Starter.fs
namespace Initialization

open AuthZI.Deploy.MicrosoftEntra.Configuration
open AuthZI.Identity.MicrosoftEntra
open AuthZI.MicrosoftEntra
open AuthZI.MicrosoftOrleans.MicrosoftEntra
open AuthZI.Security
open AuthZI.Security.Authorization
open AuthZI.Tests.MicrosoftOrleans.Grains
open AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra
open AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common
open Microsoft.Extensions.DependencyInjection
open Microsoft.Identity.Client
open Orleans
open RootConfiguration
open System
open System.Text.Json
open Xunit.Abstractions
open Xunit.Sdk
open AuthZI.MicrosoftOrleans.Authorization

[<assembly: Orleans.ApplicationPartAttribute("AuthZI.Tests.MicrosoftOrleans.Grains")>]
()

type Starter(messageSink: IMessageSink) =
  inherit XunitTestFramework(messageSink)

  do
    // Read the configuration.
    let mutable microsoftEntraIdCredentialsJson =
      Environment.GetEnvironmentVariable("microsoftEntraExternalIdCredentials")

    if String.IsNullOrWhiteSpace(microsoftEntraIdCredentialsJson) then
      microsoftEntraIdCredentialsJson <- Literals.microsoftEntraExternalIDCredentialsJson

    let credentials = JsonSerializer.Deserialize<MicrosoftEntraCredentials>(microsoftEntraIdCredentialsJson)

    // Initialize the test data.
    let web1ClientApp =
      MicrosoftEntraExternalIDApp(
        credentials.DirectoryId,
        credentials.WebClient1.Id,
        credentials.WebClient1.Secret,
        credentials.WebClient1.AllowedScopes,
        AadAuthorityAudience.AzureAdMyOrg
      )
      
    let web2ClientApp =
      MicrosoftEntraExternalIDApp(
        credentials.DirectoryId,
        credentials.WebClient2.Id,
        credentials.WebClient2.Secret,
        credentials.WebClient2.AllowedScopes,
        AadAuthorityAudience.AzureAdMyOrg
      )

    TestData.UserWithScopeAdeleV <-
      [ [| credentials.AdeleV.Name; credentials.AdeleV.Password; [ "Api1"; "Orleans" ] |] ]

    TestData.UserWithScopeAlexW <- [ [| credentials.AlexW.Name; credentials.AlexW.Password; [ "Api1"; "Orleans" ] |] ]
    TestData.Users <- [ [| credentials.AdeleV.Name; credentials.AdeleV.Password |] ]
    TestData.Web1ClientApp <- web1ClientApp
    TestData.Web2ClientApp <- web2ClientApp

    TestData.GetAccessTokenForUserOnMicrosoftEntraAppAsync <-
      AccessTokenRetriever.getTokenByUserNameAndPasswordForEntraExternalIDTenant

    let configureSiloHost =
      fun (services: IServiceCollection) ->
        // Add Azure Active Directory authorization.
        services.AddOrleansAuthorization(
          TestData.Web1ClientApp,
          (fun (config: AuthZI.Security.Configuration) ->
            config.ConfigureAuthorizationOptions <- Action<AuthorizationOptions>(AuthorizationConfig.ConfigureOptions)),
          AuthorizationConfiguration(false)
        )
        // Some custom authorization services.
        AuthorizationConfig.ConfigureServices(services)

    // Initialize Orleans test cluster.
    let siloHost = SiloHostBuilder.Build(configureSiloHost)
    siloHost.StartAsync().Wait()

    // Initialize the SiloClient.
    let configureSiloClient =
      fun (services: IServiceCollection) ->
        let configureCluster (config: Configuration) =
          config.ConfigureAuthorizationOptions <- Action<AuthorizationOptions>(AuthorizationConfig.ConfigureOptions)

        AuthorizationConfig.ConfigureServices(services)
        services.AddSingleton<IAccessTokenProvider>(fun _ -> accessTokenProvider) |> ignore
        // Add Azure Active Directory authorization.
        services.AddOrleansClientAuthorization(TestData.Web1ClientApp, fun config -> configureCluster (config))
        |> ignore

    let siloClientHost = SiloClientBuilder.Build(configureSiloClient)
    siloClientHost.StartAsync().Wait()
    TestData.IClusterClient <- siloClientHost.Services.GetService<IClusterClient>()

//siloClientProvider.SiloClient <- SiloClient.getClusterClient()

module CurrentAssembly =
  [<Literal>]
  let TypeName = "Initialization.Starter"

  [<Literal>]
  let Name = "AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraExternalID.NET9.0"

[<assembly: Xunit.TestFramework(CurrentAssembly.TypeName, CurrentAssembly.Name)>]
()

File: test/AuthZI.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraExternalID.CSharp/Initialization/Starter.cs
using AuthZI.Deploy.MicrosoftEntra.Configuration;
using AuthZI.Identity.MicrosoftEntra;
using AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common.Initialization;
using Microsoft.Identity.Client;
using Orleans;
using Xunit;

[assembly: ApplicationPart("AuthZI.Tests.MicrosoftOrleans.Grains")]
[assembly: AssemblyFixture(
  typeof(AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraExternalID.CSharp.Initialization.ExternalIdMainTestFixture))]
[assembly: CollectionBehavior(DisableTestParallelization = true)]

namespace AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraExternalID.CSharp.Initialization;

public class ExternalIdMainTestFixture : MainTestFixture
{
  public ExternalIdMainTestFixture()
    : base(
      "microsoftEntraExternalIdCredentials",
      Literals.microsoftEntraExternalIDCredentialsJson,
      "microsoftEntraExternalIdRefreshTokens",
      "microsoftEntraExternalIdRefreshTokensPath",
      RefreshTokenStore.MicrosoftEntraExternalID1)
  {
  }

  protected override MicrosoftEntraApp CreateMicrosoftEntraApp(
    MicrosoftEntraCredentials credentials,
    Client client) =>
    new MicrosoftEntraExternalIDApp(
      credentials.DirectoryId,
      client.Id,
      client.Secret,
      client.AllowedScopes,
      AadAuthorityAudience.AzureAdMyOrg);

  protected override string GetTokenByUserNameAndPassword(
    MicrosoftEntraApp entraIdApp,
    string userName,
    string password) =>
    AccessTokenRetriever.GetTokenByUserNameAndPasswordForEntraExternalIdTenant(entraIdApp, userName, password);

  protected override string GetTokenByRefreshToken(
    MicrosoftEntraApp entraIdApp,
    string userName,
    MicrosoftEntraRefreshTokenStore refreshTokenStore) =>
    AccessTokenRetriever.GetTokenByRefreshTokenForEntraExternalIdTenant(entraIdApp, userName, refreshTokenStore);
}
